### PR TITLE
expose remote document resolver function

### PR DESCRIPTION
**Pull request by FedeBev**
## The why
This PR would resolve issue #132 
At date, remote file are downloaded with a [simple http client](https://github.com/pb33f/libopenapi/blob/f7901747e8812b4f4e2c454942b487b067c987d2/index/find_component.go#L88-L103). This doesn't allow to dowload files from sources that require a more "complex" client, such as authentication.
An example is a reference to a private github repo

## The solution

[`datamodel.DocumentConfiguration`](https://github.com/pb33f/libopenapi/blob/f7901747e8812b4f4e2c454942b487b067c987d2/datamodel/document_config.go#L13) exposes a new field: `RemoteDocumentGetter func(url string) (*http.Response, error)`
`RemoteDocumentGetter` is propagated to [`index.SpecIndexConfig`](https://github.com/pb33f/libopenapi/blob/f7901747e8812b4f4e2c454942b487b067c987d2/index/index_model.go#L48 
) and, when set, used intead of the simple `http.Client{}.Get` (currently in use) 

## Working example

```go
func remoteURLHandler(url string) (*http.Response, error) {
	c := http.Client{}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		zap.L().Error("cannot create request", zap.Error(err))
		return nil, err
	}

	// Accept: application/vnd.github.v3.raw
	req.Header.Add("Accept", "application/vnd.github.v3.raw")

	// Authorization: token GITHUB_TOKEN
	req.Header.Add("Authorization", fmt.Sprintf("token %s", os.Getenv("GITHUB_TOKEN")))

	resp, err := c.Do(req)
	if err != nil {
		zap.L().Error("cannot make request", zap.Error(err))
	}

	return resp, err
}

func main() {

	// load in the petstore sample OpenAPI specification
	// into a byte slice.
	petstore, _ := os.ReadFile("petstore.yaml")

	// create a new Document from from the byte slice.
	document, err := libopenapi.NewDocumentWithConfiguration(petstore, &datamodel.DocumentConfiguration{
		AllowRemoteReferences: true,
		RemoteURLHandler:  remoteURLHandler,
	})

}
```

## Bonus
I removed a bunch of unused fields, if this refactor is not welcome I'll rollback

### CODE REVIEW

File: index/find_component.go [hunk 1]
```diff
@@ -87,8 +88,10 @@ func (index *SpecIndex) FindComponent(componentId string, parent *yaml.Node) *Re
 
 var httpClient = &http.Client{Timeout: time.Duration(60) * time.Second}
 
-func getRemoteDoc(u string, d chan []byte, e chan error) {
-	resp, err := httpClient.Get(u)
+type RemoteDocumentGetter = func(url string) (*http.Response, error)
```
Comment (daveshanley) [inline index/find_component.go:91]: `RemoteURLHandler`

File: index/index_model.go [hunk 2]
```diff
@@ -60,6 +61,10 @@ type SpecIndexConfig struct {
 	// More details on relative references can be found in issue #73: https://github.com/pb33f/libopenapi/issues/73
 	BaseURL *url.URL // set the Base URL for resolving relative references if the spec is exploded.
 
+	// If resolving remotely, the RemoteDocumentGetter will be used to fetch the remote document.
+	// If not set, the default http client will be used.
+	RemoteDocumentGetter func(url string) (*http.Response, error)
```
Comment (daveshanley) [inline index/index_model.go:66]: Add a link to the issue please, so others can follow the story.

File: index/index_model.go [hunk 3]
```diff
@@ -161,14 +166,8 @@ type SpecIndex struct {
 	operationTagsCount                  int                                           // number of unique tags in operations
 	globalTagsCount                     int                                           // number of global tags defined
 	totalTagsCount                      int                                           // number unique tags in spec
-	securitySchemesCount                int                                           // security schemes
```
[the remark this answers is not in the document]
Comment (FedeBev) [reply index/index_model.go:164]: I can definitely do that if you are uncomfortable, however keep in mind that
- all those fields are private, therefore no impact on user
- code builds, tests are passing (just a confirmation that where unused)
- staticcheck linter confirms that are unused

I don't see any warning sign to be honest, let me know what you think

File: index/index_model.go (hunk 2 above)
Comment (FedeBev) [reply index/index_model.go:66]: will do

File: index/find_component.go (hunk 1 above)
Comment (FedeBev) [reply index/find_component.go:91]: will do

File: index/find_component.go [hunk 4]
```diff
@@ -87,8 +88,10 @@ func (index *SpecIndex) FindComponent(componentId string, parent *yaml.Node) *Re
...
+
+func getRemoteDoc(g RemoteDocumentGetter, u string, d chan []byte, e chan error) {
```
[the remark this answers is not in the document]
Comment (FedeBev) [reply index/find_component.go:93]: definitely, will do

File: index/find_component.go [hunk 5]
```diff
@@ -121,7 +124,11 @@ func (index *SpecIndex) lookupRemoteReference(ref string) (*yaml.Node, *yaml.Nod
 		go func(uri string) {
 			bc := make(chan []byte)
 			ec := make(chan error)
-			go getRemoteDoc(uri, bc, ec)
+			var getter RemoteDocumentGetter = httpClient.Get
+			if index.config != nil && index.config.RemoteDocumentGetter != nil {
+				getter = index.config.RemoteDocumentGetter
```
[the remark this answers is not in the document]
Comment (FedeBev) [reply index/find_component.go:129]: definitely, will do

File: datamodel/document_config.go [hunk 6]
```diff
@@ -15,6 +18,10 @@ type DocumentConfiguration struct {
 	// Schema must be set to "http/https".
 	BaseURL *url.URL
 
+	// RemoteDocumentGetter is a function that will be used to retrieve remote documents. If not set, the default
+	// remote document getter will be used.
+	RemoteDocumentGetter func(url string) (*http.Response, error)
```
[the remark this answers is not in the document]
Comment (FedeBev) [reply datamodel/document_config.go:23]: will do

File: index/index_model.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (daveshanley) [reply index/index_model.go:164]: Call it superstition or just an overwhelming abundance of caution. I hear you - I am just highly cautious about deleting things I *know* I put in there for a reason and just have not got around to thoroughly evaluating if it *really* needs to be there or should just be removed.

**Review by daveshanley — APPROVED** [review]
Looks good to me! Thank you for the upgrade!

**Pull request merged**